Semantics.fs

```diff
@@ -442,193 +436,239 @@ let tliftToMicrocode
+            updateState (updateState s t) e
+    List.fold updateOne state listing
 
 /// <summary>
 ///     Converts a microcode instruction set into a two-state Boolean predicate.
 /// </summary>
-let microcodeToBool
-  (instrs : Microcode<Expr<Sym<Var>>, Sym<Var>> list)
-  : Result<BoolExpr<Sym<MarkedVar>>, Error> =
-    (* First, normalise the expression: this will make framing much easier
-       later. *)
-    let normalisedResult = normaliseMicrocode instrs
-
-    (* We need some traversals to rewrite the expressions to their
-       pre- and post-states.  The normalisation has already sifted things that
-       are post-states onto the LHS of an Assign, and pre-states everywhere
-       else. *)
-    let toAfter = traverseTypedSymWithMarker After
-    let toAfterV =
-        mapTraversal (tliftToExprDest toAfter)
-        >> mapMessages Traversal
-    let toBefore = traverseTypedSymWithMarker Before
-    let toBeforeE =
-        mapTraversal (tliftOverExpr toBefore) >> mapMessages Traversal
-    let toBeforeB =
-        mapTraversal (tliftToBoolSrc (tliftToExprDest toBefore))
-        >> mapMessages Traversal
-
-    let rec translateInstrs is =
-        // TODO(CaptainHayashi): convert to variable LVs.
-        // TODO(CaptainHayashi): framing (currently done elsewhere).
-        let translateInstr =
-            function
-            | Assign (x, y) -> lift2 mkEq (toAfterV x) (toBeforeE y)
-            | Assume x -> toBeforeB x
-            | Branch (i, t, e) ->
-                lift3
-                    (fun iX tX eX ->
-                        mkAnd2 (mkImplies iX tX) (mkImplies (mkNot iX) eX))
-                    (toBeforeB i)
-                    (translateInstrs t)
-                    (translateInstrs e)
-        lift mkAnd (collect (List.map translateInstr is))
-
-    bind translateInstrs normalisedResult
-
-/// Instantiates a PrimCommand from a PrimSemantics instance
-let instantiatePrim
-  (smap : PrimSemanticsMap)
-  (prim : PrimCommand)
-  : Result<SMBoolExpr option, Error> =
-    let primDefResult = lookupPrim prim smap
-    let typeCheckedDefResult =
-        bind
-            (function
-             | None -> ok None
-             | Some sem ->
-                lift (fun _ -> Some sem) (checkParamTypesPrim prim sem))
-            primDefResult
-
-    let instantiate =
-        function
-        | None -> ok None
-        | Some s ->
-            let subInMCode =
-                    tchainL (tliftToMicrocode (primParamSubFun prim s)) id
-            let subbedMCode =
-                mapMessages Traversal (mapTraversal subInMCode s.Body)
-            let subbed = bind microcodeToBool subbedMCode
-            lift Some subbed
-
-    bind instantiate typeCheckedDefResult
-
-/// Translate a primitive command to an expression characterising it
-/// by instantiating the semantics from the core semantics map with
-/// the variables from the command
-let instantiateSemanticsOfPrim
+let rec markedMicrocodeToBool
```

Comment:
A lot of the preprocessing for `Microcode` used to be in `microcodeToBool`, and has now been split out into other functions.  This means `markedMicrocodeToBool` literally just does the final leg in converting processed `Microcode` to a `BoolExpr`.